File: app/src/main/java/com/b502/minedroid/GameActivity.java
package com.b502.minedroid;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatButton;

import com.b502.minedroid.utils.MapManager;

public class GameActivity extends AppCompatActivity implements View.OnClickListener {

    private MapManager mapManager;
    MapManager.GameDifficulty dif;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game);

        Intent i = getIntent();
        dif = MapManager.GameDifficulty.values()[i.getIntExtra("diff", 0)];
        AppCompatButton btnSmile = findViewById(R.id.btnsmile);
        btnSmile.setOnClickListener(this);
        mapManager = new MapManager(this, dif);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onClick(View view) {
        if (mapManager != null) {
            mapManager.reset();
        }
    }

    @Override
    protected void onPause() {
        mapManager.getTimeManagementMaster().pause();
        super.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mapManager.getTimeManagementMaster().isHangedup()) {
            mapManager.getTimeManagementMaster().start();
        }
    }

    @Override
    protected void onDestroy() {
        mapManager.getTimeManagementMaster().stop();
        super.onDestroy();
    }
}

File: app/src/main/java/com/b502/minedroid/GameActivity.kt
package com.b502.minedroid

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import com.b502.minedroid.utils.MapManager
import com.b502.minedroid.utils.MapManager.Difficulty

class GameActivity : AppCompatActivity() {
    private lateinit var difficulty: Difficulty
    private lateinit var mapManager: MapManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game)
        val btnSmile = findViewById<AppCompatButton>(R.id.btnsmile)
        btnSmile.setOnClickListener { mapManager.reset() }
        difficulty = Difficulty.entries.toTypedArray()[intent.getIntExtra("diff", 0)]
        mapManager = MapManager(this, difficulty)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onPause() {
        mapManager.timer.pause()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        mapManager.timer.resume()
    }

    override fun onDestroy() {
        mapManager.timer.stop()
        super.onDestroy()
    }
}